module/component/component.go

```diff
@@ -0,0 +1,320 @@
+package component
+
+import (
+	"context"
+	"errors"
+	"sync"
+
+	"go.uber.org/atomic"
+	"golang.org/x/sync/errgroup"
+
+	"github.com/onflow/flow-go/module"
+	"github.com/onflow/flow-go/module/irrecoverable"
+	"github.com/onflow/flow-go/module/util"
+)
+
+var ErrMultipleStartup = errors.New("component may only be started once")
+
+type Component interface {
+	module.Startable
+	module.ReadyDoneAware
+}
+
+type NoopComponent struct{}
+
+var _ Component = (*NoopComponent)(nil)
+
+func (c *NoopComponent) Start(irrecoverable.SignalerContext) error { return nil }
+func (c *NoopComponent) Ready() <-chan struct{}                    { return nil }
+func (c *NoopComponent) Done() <-chan struct{}                     { return nil }
+
+type ComponentFactory func() (Component, error)
+
+// OnError reacts to an irrecoverable error
+// It is meant to inspect the error, determining its type and seeing if e.g. a restart or some other measure is suitable,
+// and optionally trigger the continuation provided by the caller (RunComponent), which defines what "a restart" means.
+// Instead of restarting the component, it could also:
+// - panic (in canary / benchmark)
+// - log in various Error channels and / or send telemetry ...
+type OnError = func(err error, triggerRestart func())
+
+func RunComponent(ctx context.Context, componentFactory ComponentFactory, handler OnError) error {
+	// reference to per-run signals for the component
+	var component Component
+	var cancel context.CancelFunc
+	var done <-chan struct{}
+	var irrecoverableErr <-chan error
+
+	start := func() (err error) {
+		component, err = componentFactory()
+		if err != nil {
+			return // failure to generate the component, should be handled out-of-band because a restart won't help
+		}
+
+		// context used to run the component
+		var runCtx context.Context
+		runCtx, cancel = context.WithCancel(ctx)
+
+		// signaler used for irrecoverables
+		signaler := irrecoverable.NewSignaler()
+		signalingCtx := irrecoverable.WithSignaler(runCtx, signaler)
+		irrecoverableErr = signaler.Error()
+
+		// the component must be started in a separate goroutine in case an irrecoverable error
```

Comment:
is this still true now that we're explicitly returning an error?

Reply:
i.e. is the convention that implementors MUST NOT throw an irrecoverable during Start and should return an error instead since subroutines would all be started in their own goroutine anyway.

Reply:
I left it there just in case someone misuses the interface, but you could argue it might just be better to document more clearly and assume that developers won't make the mistake.